File: IsometricRPG/IsometricRPG/Combat/CombatSystem.swift
import SpriteKit

/// Manages all active bullets and handles hit detection.
final class CombatSystem {
    private(set) var bullets: [Bullet] = []
    private weak var worldNode: SKNode?

    init(worldNode: SKNode) {
        self.worldNode = worldNode
    }

    func fireBullet(from origin: CGPoint, direction: CGPoint, at time: TimeInterval) {
        let bullet = Bullet(worldPosition: origin, direction: direction, spawnTime: time)
        bullets.append(bullet)
        worldNode?.addChild(bullet.node)

        // Muzzle flash
        if let worldNode = worldNode {
            let flash = SKShapeNode(circleOfRadius: 8)
            flash.fillColor = SKColor.white.withAlphaComponent(0.8)
            flash.strokeColor = .clear
            flash.position = IsometricMath.worldToScreen(origin)
            flash.zPosition = Constants.ZPosition.bullet + 1
            worldNode.addChild(flash)
            flash.run(SKAction.sequence([
                SKAction.fadeOut(withDuration: 0.08),
                SKAction.removeFromParent()
            ]))
        }
    }

    func update(deltaTime: TimeInterval, currentTime: TimeInterval, enemies: inout [Enemy], player: Player) {
        // Update bullets
        for bullet in bullets {
            bullet.update(deltaTime: deltaTime, currentTime: currentTime)
        }

        // Bullet-enemy collisions
        for bullet in bullets where !bullet.isExpired {
            for enemy in enemies where enemy.isAlive {
                let dist = IsometricMath.distance(bullet.worldPosition, enemy.worldPosition)
                if dist < 0.5 { // half a tile
                    enemy.takeDamage(bullet.damage)
                    bullet.isExpired = true

                    if !enemy.isAlive {
                        player.gainExperience(enemy.experienceValue)
                    }
                    break
                }
            }
        }

        // Clean up expired bullets
        let expired = bullets.filter { $0.isExpired }
        for bullet in expired {
            bullet.remove()
        }
        bullets.removeAll { $0.isExpired }

        // Clean up dead enemies
        enemies.removeAll { !$0.isAlive && $0.node.parent == nil }
    }

    func removeAll() {
        for bullet in bullets {
            bullet.node.removeFromParent()
        }
        bullets.removeAll()
    }
}

File: IsometricRPGMono/Combat/CombatSystem.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;

namespace IsometricRPG;

public class CombatSystem
{
    private readonly List<Bullet> _bullets = new List<Bullet>();

    public IReadOnlyList<Bullet> Bullets => _bullets;

    public void FireBullet(Vector2 fromWorldPos, Vector2 direction, double currentTime)
    {
        if (direction.LengthSquared() < 1e-6f) return;
        _bullets.Add(new Bullet(fromWorldPos, direction, currentTime, Constants.BulletDamage));
    }

    public void Update(double deltaTime, double currentTime, List<Enemy> enemies, Player player)
    {
        // Advance all bullets
        foreach (var bullet in _bullets)
            bullet.Update(deltaTime, currentTime);

        // Bullet-enemy collisions
        foreach (var bullet in _bullets)
        {
            if (bullet.IsExpired) continue;
            foreach (var enemy in enemies)
            {
                if (!enemy.IsAlive) continue;
                if (bullet.CheckHit(enemy))
                {
                    enemy.TakeDamage(bullet.Damage);
                    bullet.IsExpired = true;

                    if (!enemy.IsAlive && player != null)
                        player.AddXP(enemy.XPReward);

                    break;
                }
            }
        }

        _bullets.RemoveAll(b => b.IsExpired);
    }

    public void Draw(SpriteBatch spriteBatch, SpriteManager spriteManager, Vector2 cameraPos)
    {
        foreach (var bullet in _bullets)
            bullet.Draw(spriteBatch, spriteManager, cameraPos);
    }

    public void Clear() => _bullets.Clear();
}
